lib/imports/api/authority-registry/server/authority-registry.js

```diff
@@ -0,0 +1,319 @@
+import {
+  insertAuthority,
+  removeAuthority,
+  storeFailedHealthChecks,
+  removeAuthorities,
+  resetFailedHealthChecksCount,
+  addDocToAuthority
+} from './methods';
+import { authorityRegistryColl } from './collection';
+import { WebApp } from 'meteor/webapp';
+import { HTTP } from 'meteor/http';
+import { check } from 'meteor/check';
+import { Meteor } from 'meteor/meteor';
+import ip from 'ip';
+
+const HTTPS_PORT = 443;
+const HEALTH_CHECK_RESPONSE = 'ProseMeteor Health OK';
+const FAILED_HEALTH_CHECKS_THRESHOLD = 5;
+const HEALTH_CHECK_INTERVAL_MS = 1000 * 30;   // interval on which to run health checks
+
+/**
+*   Class that manages and communicates with the authority registry. Maintains a list of all
+*   authority servers.
+*/
+export class AuthorityRegistry {
+  constructor () {
+    this.ip = ip.address();
+    this.port = parseInt(process.env.PORT);
+
+    const ipAndPort = `${this.ip}:${this.port}`;
+    this.fullUrl = (this.port === HTTPS_PORT) ? `https://${ipAndPort}` : `http://${ipAndPort}`;
+    console.log(`ProseMeteor AuthorityRegistry, this server url: ${this.ipAndPort}`);
+
+    // bind class methods
+    this.setUpHealthCheckEndpoint = this.setUpHealthCheckEndpoint.bind(this);
+    this.setupClientMethods = this.setupClientMethods.bind(this);
+    this.startHealthCheckInterval = this.startHealthCheckInterval.bind(this);
+    this.register = this.register.bind(this);
+    this.unregister = this.unregister.bind(this);
+    this.runHealthChecks = this.runHealthChecks.bind(this);
+    this.fullRegistry = this.fullRegistry.bind(this);
+    this.isRegistered = this.isRegistered.bind(this);
+    this.getEligibleAuthoritiesToHostDoc = this.getEligibleAuthoritiesToHostDoc.bind(this);
+
+    this.setUpHealthCheckEndpoint();
+    this.startHealthCheckInterval();
+    this.setupClientMethods();
+  }
+  /**
+  * Creates a health check endpoint for this server
+  */
+  setUpHealthCheckEndpoint () {
+    // create a health check endpoint, so other servers can verify this server is alive or not
+    WebApp.rawConnectHandlers.use('/prosemeteor-health-check', (req, res, cb) => {
+      res.writeHead(200, {
+        'Content-Type': 'text/plain'
+      });
+      res.end(HEALTH_CHECK_RESPONSE);
+      cb();
+    });
+  }
+
+  /**
+  * Creates Meteor Methods to allow clients to communicate with the AuthorityRegistry.
+  */
+  setupClientMethods () {
+    // let self = this;
+    // Meteor.methods({});
+  }
+  /**
+  *  Starts a continuous interval to run health checks against all other servers so
+  *  we can know when an authority server is no longer available.
+  */
+  startHealthCheckInterval () {
```

Comment:
im having a little trouble understanding the intended  state diagram for the health checks... especially the fail paths...

if a previously available server is no longer available you retry a definable number of times and then remove it.... is the intention to automatically re-open it on some other server

unregistering looks like its a hard command but once docs can be removed from an authority should unregistering be a longer process where the server no longer accepts new docs and waits until its last doc has been closed to truly unregister... or do we migrate the docs on the server to be unregistered to a new server